lang/cs/Org.Apache.REEF.Wake/Remote/Impl/RemoteConnectionRetryHandler.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Net;
+using System.Net.Sockets;
+using Microsoft.Practices.TransientFaultHandling;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Wake.Remote.Impl
+{
+    /// <summary>
+    /// Handles the retry logic to connect to remote endpoint
+    /// </summary>
+    internal class RemoteConnectionRetryHandler
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(RemoteConnectionRetryHandler));
+        private readonly RetryPolicy<AllErrorsTransientStrategy> _retryPolicy;
+        private readonly int _retries;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="retries">Number of times to retry connection</param>
+        /// <param name="timeInterval">Time interval between retries</param>
+        internal RemoteConnectionRetryHandler(int retries, int timeInterval)
+        {
+            _retries = retries;
+            _retryPolicy = new RetryPolicy<AllErrorsTransientStrategy>(
+                new FixedInterval("ConnectionRetries", retries, new TimeSpan(0, 0, 0, 0, timeInterval), true));
+            _retryPolicy.Retrying += (sender, args) =>
+            {
+                // Log details of the retry.
+                var msg = string.Format("Retry - Count:{0}, Delay:{1}, Exception:{2}",
+                    args.CurrentRetryCount, args.Delay, args.LastException);
+                Logger.Log(Level.Info, msg);
+            };
+        }
+
+        /// <summary>
+        /// Connects to the remote endpoint
+        /// </summary>
+        /// <param name="endPoint">Remote endpoint</param>
+        /// <returns></returns>
+        internal TcpClient Connect(IPEndPoint endPoint)
+        {
+            TcpClient client = new TcpClient();
+
+            try
+            {
+                _retryPolicy.ExecuteAction(() => client.Connect(endPoint));
+                return client;
+            }
+            catch (Exception e)
+            {
+                var msg = string.Format("Retried {0} times but failed", _retries);
+                Logger.Log(Level.Info, msg);
+                throw e;
```

Comment:
```c#
throw e;
```
is very bad pattern. It resets the exception stack to the new location. http://stackoverflow.com/a/881489
Please use 
```c#
throw;
```